### PR TITLE
WindowsBase zlib, ZipSharp and Package fixes for Windows

### PR DESCRIPTION
**Pull request by henricm**
This PR fixes failures on WindowsBase tests.

* Adding zlib files to MonoPosixHelper dll VS project.
* Making sure unsigned long are passed correctly from zlib to managed code which differs from other platforms since unsigned long is always 32 bits when using MS VS compiler.

* Setting Culture to `en-us` for `RectTest.ToStringTest`.

* Making sure `Package` stream is closed when exception is thrown and no instance was returned from `Package.Open`. Stream will never be closed if we throw exception in Open which leads to file being locked in Windows and that we leak file handles.

### CODE REVIEW

File: mcs/class/WindowsBase/WindowsBase-net_4_x.csproj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (marek-safar) [inline mcs/class/WindowsBase/WindowsBase-net_4_x.csproj]: This file is auto-generated

Comment (henricm) [inline mcs/class/WindowsBase/WindowsBase-net_4_x.csproj]: How is it generated?
Should it be regenerated and commited after that instead or ignored/removed from git?

Comment (marek-safar) [inline mcs/class/WindowsBase/WindowsBase-net_4_x.csproj]: will be done by someone else later

Comment (henricm) [inline mcs/class/WindowsBase/WindowsBase-net_4_x.csproj]: Great. I'll update the commit.

Comment (akoeplinger) [inline mcs/class/WindowsBase/WindowsBase-net_4_x.csproj]: FYI you can regenerate the .csproj files by running `make update-solution-files` in the repo root.

The reason why they're checked in is to eventually enable you to build the class libs with msbuild without needing automake etc. to generate the files first.

**Comment by ntherning** [comment]
build

**Comment by ntherning** [comment]
@marek-safar / @akoeplinger Are we happy with this PR now? I have triggered a new build to make sure it doesn't break anything against latest HEAD.

**Comment by marek-safar** [comment]
LGTM

File: mcs/class/WindowsBase/System.IO.Packaging/Package.cs [hunk 2]
```diff
@@ -404,7 +404,12 @@ public static Package Open (string path, FileMode packageMode, FileAccess packag
 				throw new FileFormatException ("Stream length cannot be zero with FileMode.Open");
 
 			Stream s = File.Open (path, packageMode, packageAccess, packageShare);
-			return Open (s, packageMode, packageAccess, true);
+		    try {
+		        return Open (s, packageMode, packageAccess, true);
+		    } catch {
+				s.Close  ();
+				throw;
+		    }
```
Comment (akoeplinger) [inline mcs/class/WindowsBase/System.IO.Packaging/Package.cs]: IMHO a more idiomatic way would be to do either

```csharp
try {
    return Open (s, packageMode, packageAccess, true);
} finally {
    s.Close ();
}
```

or 

```csharp
using (Stream s = File.Open (path, packageMode, packageAccess, packageShare))
   return Open (s, packageMode, packageAccess, true);

**Comment by akoeplinger** [comment]
There was a crash in the WindowsBase test suite on `Linux x64`, please take a look and make sure it's not related to the changes :)

**Comment by ntherning** [comment]
Yeah, we're looking into it.

File: mcs/class/WindowsBase/System.IO.Packaging/Package.cs (hunk 2 above)
Comment (marek-safar) [inline mcs/class/WindowsBase/System.IO.Packaging/Package.cs]: My reading is that'd be wrong as the stream is passed on on success

Comment (akoeplinger) [inline mcs/class/WindowsBase/System.IO.Packaging/Package.cs]: Ohhh, ok, I missed that. Makes sense 😄

File: mcs/class/WindowsBase/WindowsBase-net_4_x.csproj (hunk 1 above)
Comment (henricm) [inline mcs/class/WindowsBase/WindowsBase-net_4_x.csproj]: Thanks for the info!

**Comment by ntherning** [comment]
@akoeplinger This PR has been updated to do the platform specific decisions at runtime. It's not the prettiest but it seems to do the job. :-) I have gone through the latest PR builds and none of the test failures are related to this change. Please have a look and comment if needed. Thanks!

File: mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs [hunk 3]
```diff
@@ -30,6 +30,7 @@ namespace zipsharp
 {
 	class UnzipReadStream : Stream
 	{
+		private static bool use32Bit = IntPtr.Size == 4 || Environment.OSVersion.Platform != PlatformID.Unix;
```
Comment (marek-safar) [inline mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs]: Could you move all copies into a new class e.g. `static class NativeUnzipVersion` and add comment why it's done like that

Comment (akoeplinger) [inline mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs]: You could also just do all this 32/64 bit logic inside NativeUnzip then you don't need to modify all the callers.

Comment (marek-safar) [inline mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs]: I am not sure I'd split the logic as some methods have different parameter type and other don't

Comment (henricm) [inline mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs]: @marek-safar Maybe just put the `use32Bit` field in IOFunctions? It's used for both zip and unzip...

Comment (henricm) [inline mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs]: ah, forget that. Just realised that wasn't a class...

Comment (henricm) [inline mcs/class/WindowsBase/ZipSharp/UnzipReadStream.cs]: Added class `NativeVersion` which holds the Use32Bit constant.

File: mcs/class/WindowsBase/ZipSharp/NativeUnzip.cs [hunk 4]
```diff
@@ -178,15 +205,25 @@ public static unsafe int Read (UnzipHandle handle, byte[] buffer, int offset, in
...
+		                                            uint extraFieldBufferSize, // uLong
+		                                            StringBuilder szComment,
+		                                            uint commentBufferSize);   // uLong
+
+		[DllImport ("MonoPosixHelper", EntryPoint="unzGetCurrentFileInfo", CallingConvention=CallingConvention.Cdecl)]
+		static extern int unzGetCurrentFileInfo_64 (UnzipHandle handle,
+		                                            out UnzipFileInfo64 pfile_info,
+		                                            StringBuilder szFileName,
+		                                            ulong fileNameBufferSize,   // uLong
+		                                            IntPtr extraField,           // void *
```
Comment (akoeplinger) [inline mcs/class/WindowsBase/ZipSharp/NativeUnzip.cs]: nit: make the comments align

File: mcs/class/WindowsBase/ZipSharp/UnzipFileInfo.cs [hunk 5]
```diff
@@ -10,103 +10,46 @@
...
-	    IntPtr flag;                 /* general purpose bit flag        2 bytes */
-	    IntPtr compression_method;   /* compression method              2 bytes */
-	    IntPtr dosDate;              /* last mod file date in Dos fmt   4 bytes */
-	    IntPtr crc;                  /* crc-32                          4 bytes */
-	    IntPtr compressed_size;      /* compressed size                 4 bytes */
-	    IntPtr uncompressed_size;    /* uncompressed size               4 bytes */
-	    IntPtr size_filename;        /* filename length                 2 bytes */
-	    IntPtr size_file_extra;      /* extra field length              2 bytes */
-	    IntPtr size_file_comment;    /* file comment length             2 bytes */
+	    public uint version;              /* version made by                 2 bytes */
```
Comment (akoeplinger) [inline mcs/class/WindowsBase/ZipSharp/UnzipFileInfo.cs]: mixed tabs/spaces

File: mcs/class/WindowsBase/ZipSharp/NativeVersion.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+		/// gcc/clang the C long type follows the bitness of the targeted architecture,
+		/// it's 32-bit on 32-bit systems and 64-bit on 64-bit systems. With the VS
+		/// compiler however, the C long type is always 32-bit regardless of the
+		/// target architecture. zlib and minizip uses C long in a number of 
+		/// different function signatures and structs. 
+		/// 
+		/// This field is used to easily determine if the 32 bit version of 
+		/// functions and structs should be used when interacting with zlib.
+		/// </summary>
+		public static bool Use32Bit = IntPtr.Size == 4 || Environment.OSVersion.Platform != PlatformID.Unix;
```
Comment (akoeplinger) [inline mcs/class/WindowsBase/ZipSharp/NativeVersion.cs]: nit: could be `readonly`

Comment (henricm) [inline mcs/class/WindowsBase/ZipSharp/NativeVersion.cs]: Above fixed in latest push.

**Comment by henricm** [comment]
Above fixed in latest push.

**Pull request merged**